aip/general/0129.md

```diff
@@ -0,0 +1,65 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-modified Values and Defaults
+
+### State-driven clients (aka Convergent clients)
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
+When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An API service ***must*** treat a client-side dynamic default field as read-only and must provide a separate `OUTPUT_ONLY` field that reflects the server assigned, allocated, or generated default value. 
```

Comment:
Should "read-only" be "write-only" here, or am I misunderstanding something? _Ideally_ they'd be read-write but never canonicalized, I think, and could potentially represent a constraint, but write-only works.

---

When I say a constraint, I mean something like expressing a CIDR in the example below, and having the service allocate an ip in that range. For example:

input:

```
IpAddress = "192.168.1.0/24"
```

output:

```
IpAddress = "192.168.1.0/24"
ProvisionedIpAddress = "192.168.1.7"
```

Reply:
The text was intending to say the service shouldn't populate empty mutable fields or change values provided by the client.

Reply:
Oh, got it, read-only from an implementation of a service perspective and not from the frontend of the API perspective. That makes more sense, thanks!

I think this could use a little more front matter to explain the approach. Something along the lines of "An attribute with a dynamic default **must** be expressed with two fields in the API, a mutable field that can optionally be set by the user, and an output-only field that records the final value decided on by the API service."

Reply:
Added wording specifying two separate fields

Reply:
Works for me, thanks!